src/sinks/util/auto_concurrency/controller.rs

```diff
@@ -0,0 +1,243 @@
+    }
+
+    pub(super) fn start_request(&self) {
+        let mut inner = self.inner.lock().expect("Controller mutex is poisoned");
+
+        #[cfg(test)]
+        {
+            let mut stats = self.stats.lock().expect("Stats mutex is poisoned");
+            stats.in_flight.add(inner.in_flight, instant_now());
+        }
+
+        inner.in_flight += 1;
+        emit!(AutoConcurrencyInFlight {
+            in_flight: inner.in_flight as u64
+        });
+    }
+
+    fn adjust_to_back_pressure(&self, start: Instant, is_back_pressure: bool) {
+        let now = instant_now();
+        let mut inner = self.inner.lock().expect("Controller mutex is poisoned");
+
+        #[cfg(test)]
+        let mut stats = self.stats.lock().expect("Stats mutex is poisoned");
+
+        let rtt = now.saturating_duration_since(start);
+        emit!(AutoConcurrencyObservedRtt { rtt });
+        let rtt = rtt.as_secs_f64();
+
+        #[cfg(test)]
+        stats.observed_rtt.add(rtt, now);
+
+        if is_back_pressure {
+            inner.had_back_pressure = true;
+        }
+
+        #[cfg(test)]
+        stats.in_flight.add(inner.in_flight, now);
+
+        inner.in_flight -= 1;
+        emit!(AutoConcurrencyInFlight {
+            in_flight: inner.in_flight as u64
+        });
+
+        let rtt = inner.current_rtt.update(rtt);
+        let avg = inner.past_rtt.average();
+
+        if avg == 0.0 {
+            // No past measurements, set up initial values.
+            inner.past_rtt.update(rtt);
+            inner.next_update = now + Duration::from_secs_f64(rtt);
+        } else if avg > 0.0 && now >= inner.next_update {
+            emit!(AutoConcurrencyAveragedRtt {
+                rtt: Duration::from_secs_f64(rtt)
+            });
+
+            #[cfg(test)]
+            {
+                stats.averaged_rtt.add(rtt, now);
+                stats.concurrency_limit.add(inner.current_limit, now);
+            }
+
+            let threshold = avg * THRESHOLD_RATIO;
+
+            // Back pressure responses, either explicit or implicit due
+            // to increasing response times, trigger a decrease in the
+            // concurrency limit.
+            if inner.current_limit > 1 && (inner.had_back_pressure || rtt >= avg + threshold) {
+                // Decrease (multiplicative) the current concurrency limit
+                let to_forget = inner.current_limit / 2;
+                self.semaphore.forget_permits(to_forget);
+                inner.current_limit -= to_forget;
+            }
+            // Normal quick responses trigger an increase in the
+            // concurrency limit. Note that we only check this if we had
+            // requests to go beyond the current limit to prevent
+            // increasing the limit beyond what we have evidence for.
+            else if inner.current_limit < self.max
+                && inner.reached_limit
+                && !inner.had_back_pressure
+                && rtt <= avg
```

Comment:
This `rtt <= avg` strikes me as a little bit conservative and I'm curious what the effect would be if we gave it some wiggle room (e.g. within a 2-5% of the current average or below).

Very much out of scope for this PR, but it'd be interesting to have a little simulator to test things out like that.

Reply:
Actually, this is a little big aggressive, as this is the logic for _increasing_ the concurrency limit. This was the purpose of `THRESHOLD_RATIO`, but I only used it the other case. I went back and forth on this, so I'm open to tweaking the threshold logic.

I definitely agree it would be useful to have a simulation setup to test this. One of my thoughts was to work in both `EWMA_ALPHA` and `THRESHOLD_RATIO` as undocumented global parameters so we could test using an actual vector build, but I didn't work that into this (already large) PR.